### ISSUE TITLE
Handling of native packages

### ISSUE THREAD

**Issue by pmderodat** [alire-project/alire#65]
As of today, the handling of native packages only works on Debian or Ubuntu distributions. For instance, on my Arch Linux box, I’m getting a lot of crashes such as:
```
$ bin/alr search --full --list
Searching... .                       
raised PROGRAM_ERROR : Unsupported package manager
[/home/pmderodat/_git/alr/bin/alr]
Alr.Origins.New_Origin at alr-origins.adb:45
Alr.Query.Is_Available.Localalias at alr-query.adb:119
Alr.Query.Resolve.Expand.Expand_Value.Check at alr-query.adb:326
Alr.Query.Resolve.Expand.Expand_Value at alr-query.adb:353
Alr.Query.Resolve.Expand at alr-query.adb:408
Alr.Query.Resolve at alr-query.adb:428
Alr.Query.Is_Resolvable at alr-query.adb:126
Alr.Commands.Search.Execute.List_Release at alr-commands-search.adb:49
Alr.Commands.Search.Execute at alr-commands-search.adb:96
Alr.Commands.Execute_By_Name at alr-commands.adb:555
Alr.Commands.Execute at alr-commands.adb:533
Alr.Main at alr-main.adb:24
Main at b__alr-main.adb:1636
[/usr/lib/libc.so.6]
0x7f90deee4221
[/home/pmderodat/_git/alr/bin/alr]
0x404ecc _start at ???
0xfffffffffffffffe
```
I don’t think that requiring users to go through all packages in the index to complete package names each time they use a new distributions is a good position. Besides, the current formalism will not apply on package manager-less systems such as Windows. And this forces users to install software through their package managers, while some may want to do manual install from sources (to have recent versions, for instance).

For these reasons, I think the current mechanism to handle native package dependencies should be replaced. What about instead: assume the dependence is available then and check it is available only when trying to install a package that depends on it, or when looking for packages that are available?

Instead of using a never-exhaustive database of distribution-specific package names (which can vary over time, by the way!), maybe we could use the same trick as configure scripts: try to build a very simple program using the dependency, demonstrating that it is installed (no matter how). We could still have, as an optional feature, alr using the incomplete database to suggest which packages to install given their distribution, though…

What do you think?

**Comment by mosteo** [comment]
Adding two links for future reference:

https://github.com/rust-lang/cargo/issues/1240
Brief discussion about the same problem in cargo.

https://crates.io/crates/libudev
A native-library crate. It seems it's a simple wrapper that delegates how to install/detect outside of cargo.

**Comment by mosteo** [comment]
Also played a bit with `opam depext`, which is a dedicated command that calls the platform package manager to install whatever is needed by an opam package. Has an interesting discussion too:
https://github.com/ocaml/opam-depext
https://github.com/ocaml/opam/blob/master/doc/design/depexts-plugins

**Comment by mosteo** [comment]
I did my test with `opam install lambda-term` in a ubuntu:latest docker, that depends on native `m4` package.

The opam approach brings an interesting point which is having a separate, per platform, source of equivalences between opam names and platform names, which simplifies delegation of maintenance. In essence, that allows going forward with resolution and, on user demand, check/install OS packages needed for a particular project (that we could extend to deferring to the user the responsibility of custom installation of something).

For our use case, that might be a way of knowing that a dependency is needed but defer how to install it/what versions are available, both in the sense of alire index population and user experience. It also fits well with the "relaxed" mode we talked previously (since, worst case, we could do nothing).

I will try to come up with a proposal that integrates all the ideas we have been discussing in the near future.

**Comment by mosteo** [comment]
After #204 and #222 the original reason for this issue should be addressed, but the discussion here went further, into how to sanely use a platform's native packages and their package managers. That remains an open issue, hence the issue name change.

**Comment by mosteo** [comment]
Fixed in #303 #319
